gnome-initial-setup/pages/site/gis-site-page.c

```diff
@@ -0,0 +1,493 @@
+static gboolean
+page_validate (GisSitePage *page)
+{
+  GisSitePagePrivate *priv = gis_site_page_get_instance_private (page);
+
+  if (gtk_entry_get_text_length (GTK_ENTRY (priv->id_entry)) ||
+      gtk_entry_get_text_length (GTK_ENTRY (priv->facility_entry)) ||
+      gtk_entry_get_text_length (GTK_ENTRY (priv->street_entry)) ||
+      gtk_entry_get_text_length (GTK_ENTRY (priv->locality_entry)) ||
+      gtk_entry_get_text_length (GTK_ENTRY (priv->region_entry)) ||
+      gtk_entry_get_text_length (GTK_ENTRY (priv->country_entry)))
+    {
+      return TRUE;
+    }
+
+  return FALSE;
+}
+
+static void
+update_page_validation (GObject *object, GParamSpec *param, GisSitePage *page)
+{
+  gis_page_set_complete (GIS_PAGE (page), page_validate (page));
+}
+
+static GisSite*
+get_site_from_widgets (GisSitePage *page)
+{
+  GisSitePagePrivate *priv = gis_site_page_get_instance_private (page);
+  const gchar *id = gtk_entry_get_text (GTK_ENTRY (priv->id_entry));
+  const gchar *facility = gtk_entry_get_text (GTK_ENTRY (priv->facility_entry));
+  const gchar *street = gtk_entry_get_text (GTK_ENTRY (priv->street_entry));
+  const gchar *locality = gtk_entry_get_text (GTK_ENTRY (priv->locality_entry));
+  const gchar *region = gtk_entry_get_text (GTK_ENTRY (priv->region_entry));
+  const gchar *country = gtk_entry_get_text (GTK_ENTRY (priv->country_entry));
+
+  return gis_site_new (id, facility, street, locality, region, country);
+}
+
+static void
+ini_string_append_safe (GString     *string,
+                        const gchar *property,
+                        const gchar *value)
+{
+  g_string_append_printf (string, "%s = %s\n", property, value ? value : "");
+}
+
+static gchar*
+ini_file_content_from_site (GisSite *site)
+{
+  GString *builder = g_string_new ("[Label]\n");
+
+  /* Note the field names aren't 1:1 between GisSite and the file format */
+  ini_string_append_safe (builder, "id", gis_site_get_id (site));
+  ini_string_append_safe (builder, "facility", gis_site_get_facility (site));
+  ini_string_append_safe (builder, "street", gis_site_get_street (site));
+  ini_string_append_safe (builder, "city", gis_site_get_locality (site));
+  ini_string_append_safe (builder, "state", gis_site_get_region (site));
+  ini_string_append_safe (builder, "country", gis_site_get_country (site));
+
+  return g_string_free (builder, FALSE);
+}
+
+static void
+gis_site_page_save_data (GisPage *gis_page)
+{
+  GisSitePage *page = GIS_SITE_PAGE (gis_page);
+  gchar *location_file_content;
+  gchar *filename_created = NULL;
+  GError *error_open = NULL;
+  GError *error_close = NULL;
+  GError *error_write = NULL;
+  GError *error_pkexec = NULL;
+  GisSite *site;
+  int fd;
+
+  if (gis_page->driver == NULL)
+    return;
+
+  site = get_site_from_widgets (page);
+  location_file_content = ini_file_content_from_site (site);
+
+  fd = g_file_open_tmp (LOCATION_FILENAME_TEMPL, &filename_created,
+                        &error_open);
+  if (error_open == NULL)
```

Comment:
This block now would really benefit from an `out` section at the end of the function to free memory. So you could do
```
fd = g_file_open_tmp(&error);
if (error != NULL)
  goto out;

if (!g_close(fd, &error))
  goto out;
```

etc...